pkg/controller/registry/resolver/source_registry.go

```diff
@@ -302,3 +306,73 @@ func legacyDependenciesToProperties(dependencies []*api.Dependency) ([]*api.Prop
 	}
 	return result, nil
 }
+
+type OperatorGroupToggleSourceProvider struct {
+	sp       cache.SourceProviderWithInvalidate
+	logger   *logrus.Logger
+	ogLister v1listers.OperatorGroupLister
+}
+
+func NewOperatorGroupToggleSourceProvider(sp cache.SourceProviderWithInvalidate, logger *logrus.Logger,
+	ogLister v1listers.OperatorGroupLister) *OperatorGroupToggleSourceProvider {
+	return &OperatorGroupToggleSourceProvider{
+		sp:       sp,
+		logger:   logger,
+		ogLister: ogLister,
+	}
+}
+
+const exclusionAnnotation string = "olm.operatorframework.io/exclude-global-namespace-resolution"
+
+func (e *OperatorGroupToggleSourceProvider) Sources(namespaces ...string) map[cache.SourceKey]cache.Source {
+	// Check if annotation is set first
+	resolutionNamespaces := e.CheckForExclusion(namespaces...)
+	return e.sp.Sources(resolutionNamespaces...)
+}
+
+func (e *OperatorGroupToggleSourceProvider) Invalidate(key cache.SourceKey) {
+	e.sp.Invalidate(key)
+}
+
+func (e *OperatorGroupToggleSourceProvider) CheckForExclusion(namespaces ...string) []string {
+	var defaultResult = namespaces
+	// Only two namespaces are provided during the course of resolution, in deterministic order
+	// The first is the namespace being synced and the second is the global namespace
+	var ownNamespace = namespaces[0]
+	var globalNamespace = namespaces[1]
+	var toggledResult = []string{ownNamespace}
+
+	if ownNamespace == globalNamespace {
+		// Global namespace is being synced
+		// Return early with default
+		return defaultResult
+	}
+
+	// Check the OG on the NS provided for the exclusion annotation
+	ogs, err := e.ogLister.OperatorGroups(ownNamespace).List(labels.Everything())
+	if err != nil {
```

Comment:
That sounds reasonable -- is it best to do so in this PR? I'm just a bit wary because I think we will have to backport this down and changing the interface would expand the diff. 

I think we can get away with simply going with the default behavior in the case of an error? That does lead to the undesirable case where exclusion was specified and because the List OperatorGroup call failed we end up doing global resolution instead. That might be enough of an issue to where we would have to add an error return type to the interface in this PR?